File: src/backend/ordem_servico/src/model/os_produto_servico_model.py
from src import db
from src.model.produto_model import ProdutoModel


class OsProdutoServicoModel(db.Model):
    __tablename__ = 'os_produto_servico'

    id = db.Column(db.Integer, primary_key=True)
    tipo = db.Column(db.String(1))
    complemento = db.Column(db.Text)
    quantidade = db.Column(db.Float)
    valor_unitario = db.Column(db.Float)
    valor_subtotal = db.Column(db.Float)
    taxa_desconto = db.Column(db.Float)
    valor_desconto = db.Column(db.Float)
    valor_total = db.Column(db.Float)
    id_os_abertura = db.Column(db.Integer, db.ForeignKey('os_abertura.id'))
    id_produto = db.Column(db.Integer, db.ForeignKey('produto.id'))

    produto_model = db.relationship('ProdutoModel', foreign_keys=[id_produto])

    def mapping(self, data):
        self.id = data.get('id')
        self.id_os_abertura = data.get('idOsAbertura')
        self.id_produto = data.get('idProduto')
        self.tipo = data.get('tipo')
        self.complemento = data.get('complemento')
        self.quantidade = data.get('quantidade')
        self.valor_unitario = data.get('valorUnitario')
        self.valor_subtotal = data.get('valorSubtotal')
        self.taxa_desconto = data.get('taxaDesconto')
        self.valor_desconto = data.get('valorDesconto')
        self.valor_total = data.get('valorTotal')

    def serialize(self):
        return {
            'id': self.id,
            'idOsAbertura': self.id_os_abertura,
            'idProduto': self.id_produto,
            'tipo': self.tipo,
            'complemento': self.complemento,
            'quantidade': self.quantidade,
            'valorUnitario': self.valor_unitario,
            'valorSubtotal': self.valor_subtotal,
            'taxaDesconto': self.taxa_desconto,
            'valorDesconto': self.valor_desconto,
            'valorTotal': self.valor_total,
            'produtoModel': self.produto_model.serialize() if self.produto_model else None,
        }

File: frontend/ordem_servico/lib/app/data/model/os_produto_servico_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:ordem_servico/app/data/model/model_imports.dart';
import 'package:ordem_servico/app/data/domain/domain_imports.dart';

class OsProdutoServicoModel {
	int? id;
	int? idOsAbertura;
	int? idProduto;
	String? tipo;
	String? complemento;
	double? quantidade;
	double? valorUnitario;
	double? valorSubtotal;
	double? taxaDesconto;
	double? valorDesconto;
	double? valorTotal;
	ProdutoModel? produtoModel;

	OsProdutoServicoModel({
		this.id,
		this.idOsAbertura,
		this.idProduto,
		this.tipo,
		this.complemento,
		this.quantidade,
		this.valorUnitario,
		this.valorSubtotal,
		this.taxaDesconto,
		this.valorDesconto,
		this.valorTotal,
		this.produtoModel,
	});

	static List<String> dbColumns = <String>[
		'id',
		'tipo',
		'complemento',
		'quantidade',
		'valor_unitario',
		'valor_subtotal',
		'taxa_desconto',
		'valor_desconto',
		'valor_total',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Tipo',
		'Complemento',
		'Quantidade',
		'Valor Unitario',
		'Valor Subtotal',
		'Taxa Desconto',
		'Valor Desconto',
		'Valor Total',
	];

	OsProdutoServicoModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idOsAbertura = jsonData['idOsAbertura'];
		idProduto = jsonData['idProduto'];
		tipo = OsProdutoServicoDomain.getTipo(jsonData['tipo']);
		complemento = jsonData['complemento'];
		quantidade = jsonData['quantidade']?.toDouble();
		valorUnitario = jsonData['valorUnitario']?.toDouble();
		valorSubtotal = jsonData['valorSubtotal']?.toDouble();
		taxaDesconto = jsonData['taxaDesconto']?.toDouble();
		valorDesconto = jsonData['valorDesconto']?.toDouble();
		valorTotal = jsonData['valorTotal']?.toDouble();
		produtoModel = jsonData['produtoModel'] == null ? ProdutoModel() : ProdutoModel.fromJson(jsonData['produtoModel']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idOsAbertura'] = idOsAbertura != 0 ? idOsAbertura : null;
		jsonData['idProduto'] = idProduto != 0 ? idProduto : null;
		jsonData['tipo'] = OsProdutoServicoDomain.setTipo(tipo);
		jsonData['complemento'] = complemento;
		jsonData['quantidade'] = quantidade;
		jsonData['valorUnitario'] = valorUnitario;
		jsonData['valorSubtotal'] = valorSubtotal;
		jsonData['taxaDesconto'] = taxaDesconto;
		jsonData['valorDesconto'] = valorDesconto;
		jsonData['valorTotal'] = valorTotal;
		jsonData['produtoModel'] = produtoModel?.toJson;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idOsAbertura = plutoRow.cells['idOsAbertura']?.value;
		idProduto = plutoRow.cells['idProduto']?.value;
		tipo = plutoRow.cells['tipo']?.value != '' ? plutoRow.cells['tipo']?.value : 'Produto';
		complemento = plutoRow.cells['complemento']?.value;
		quantidade = plutoRow.cells['quantidade']?.value?.toDouble();
		valorUnitario = plutoRow.cells['valorUnitario']?.value?.toDouble();
		valorSubtotal = plutoRow.cells['valorSubtotal']?.value?.toDouble();
		taxaDesconto = plutoRow.cells['taxaDesconto']?.value?.toDouble();
		valorDesconto = plutoRow.cells['valorDesconto']?.value?.toDouble();
		valorTotal = plutoRow.cells['valorTotal']?.value?.toDouble();
		produtoModel = ProdutoModel();
		produtoModel?.nome = plutoRow.cells['produtoModel']?.value;
	}	

	OsProdutoServicoModel clone() {
		return OsProdutoServicoModel(
			id: id,
			idOsAbertura: idOsAbertura,
			idProduto: idProduto,
			tipo: tipo,
			complemento: complemento,
			quantidade: quantidade,
			valorUnitario: valorUnitario,
			valorSubtotal: valorSubtotal,
			taxaDesconto: taxaDesconto,
			valorDesconto: valorDesconto,
			valorTotal: valorTotal,
		);			
	}

	
}